### PR TITLE
add spec details popup

### PR DESCRIPTION
**Pull request by goulinkh**
## Done

- add a new API endpoint `/spec/<document-id>`: returns the document html and some metadata
- add a new view that can be showed by clicking on a spec card title

## QA

- As we don't have access to docs, you need to login
  - the app credentials: `credentials.json` (I can share this file with you)
- Start the project `dotrun`
- Login to you Google account
- Click on a spec card title
- Make sure that there is a spec details view
- Click on the close button
- Make sure that it's closed
- Click on another spec card title
- Make sure that the view is updated

### the spec details view
Untitled

### CODE REVIEW

**Comment by andesol** [comment]
@goulinkh It seems this line is throwing an error
```
self.html = BeautifulSoup(raw_html, features="lxml")
```
```
    raise FeatureNotFound(
bs4.FeatureNotFound: Couldn't find a tree builder with the features you requested: lxml. Do you need to install a parser library?
```

**Comment by goulinkh** [comment]
> @goulinkh It seems this line is throwing an error
> 
> ```
> self.html = BeautifulSoup(raw_html, features="lxml")
> ```
> 
> ```
>     raise FeatureNotFound(
> bs4.FeatureNotFound: Couldn't find a tree builder with the features you requested: lxml. Do you need to install a parser library?
> ```

Oh I must've had it locally, let me add it to the requirements, you will need to do a : `dotrun clean && dotrun install`.

File: templates/_spec_details.html [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+      </small>
+    </section>
+    <section class="spec-preview"></section>
+    <a class="p-button--positive" href="#">Open in Google Docs</a>
+  </div>
+  <script>
+    const currentElement = document.currentScript.parentElement;
+    // setup cards listener
+    const links = document.querySelectorAll(".spec-card__content a");
+    console.log("links", links);
```
Comment (andesol) [inline templates/_spec_details.html]: There are some debugging traces (console logs, prints, commented out lines). Can we get rid of them?

Comment (goulinkh) [reply templates/_spec_details.html]: :+1:

**Comment by andesol** [comment]
This is great work! A few comments:

- A few specs (e.g VisionFive image) seem to be causing on error when trying to parse its content
```
    raise ParserError("Unknown string format: %s", timestr)
dateutil.parser._parser.ParserError: Unknown string format: 2021-05-12 09:00 utc
``` 
- Could we find a way to avoid the horizontal scroll in the spec content? (maybe this can be addressed in the "design review" phase)
- The link Open in Google docs doesn't work
- Since the spec title is now triggering a UI change instead of going anywhere, it'd be better IMHO to use a button element styled as a link: https://vanillaframework.io/docs/patterns/buttons#link

File: templates/_spec_details.html [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+          specContainer.querySelector(".header").innerHTML = `
+          <li class="p-inline-list__item">${spec.metadata.index}</li>
+          <li class="p-inline-list__item">${spec.metadata.type}</li>
+        `;
+          specContainer.querySelector(".spec-preview").innerHTML = spec.html;
+          specContainer.querySelector(".title").innerHTML = spec.metadata.title;
+          specContainer.querySelector(".authors").innerHTML =
+            spec.metadata.authors.join(", ");
+          // specContainer.querySelector(".contributors").innerHTML =
+          // spec.metadata.authors.join(", ");
```
Comment (andesol) [inline templates/_spec_details.html]: Can we delete this?

**Comment by goulinkh** [comment]
> This is great work! A few comments:
> 
> * A few specs (e.g VisionFive image) seem to be causing on error when trying to parse its content
> 
> ```
>     raise ParserError("Unknown string format: %s", timestr)
> dateutil.parser._parser.ParserError: Unknown string format: 2021-05-12 09:00 utc
> ```
> 

Done :+1: 
> * Could we find a way to avoid the horizontal scroll in the spec content? (maybe this can be addressed in the "design review" phase)

I made the preview popup wider, the result seems better.
> * The link Open in Google docs doesn't work

I missed this one, well seen :+1: 
> * Since the spec title is now triggering a UI change instead of going anywhere, it'd be better IMHO to use a button element styled as a link: https://vanillaframework.io/docs/patterns/buttons#link

I intentionally left it as an anchor tag so that when the spec preview isn't working, right clicking on the title will give us the link address.
If we replace the anchor tag with a button we will not be able to access the link if the spec-preview API is down for example.

File: templates/_spec_details.html [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+          <li class="p-inline-list__item">${spec.metadata.index}</li>
+          <li class="p-inline-list__item">${spec.metadata.type}</li>
+        `;
+          specContainer.querySelector(".spec-preview").innerHTML = spec.html;
+          specContainer.querySelector(".title").innerHTML = spec.metadata.title;
+          specContainer.querySelector(".authors").innerHTML =
+            spec.metadata.authors.join(", ");
+          specContainer.querySelector(".created").innerHTML =
+            spec.metadata.created.toLocaleString();
+          specContainer.querySelector(".spec-link").href=
```
Comment (andesol) [inline templates/_spec_details.html]: I think there's something missing here and the JS is not loading

**Comment by andesol** [comment]
@goulinkh Maybe we can add `role="button"` then?

File: templates/_spec_details.html (hunk 3 above)
Comment (goulinkh) [reply templates/_spec_details.html]: Oh sorry I missed the save button before pushing :sweat_smile: Was too fast :)

**Comment by marinacastejon** [comment]
Hi @goulinkh, 
I have a couple of comments on the design. 

**1)** When the side panel is displayed, there are still 2 scrollbars (not visible directly, but when I put my cursor on the right, I can see and interact with the full-page scrollbar. 

I don't think this should be the behaviour (also could be an issue for accessibility and keyboard navigation). Instead, when the side panel is displayed, we should only be able to interact with the scrollbar of the side panel. 



**2)** The text styles don't seem to be taken from the doc and it looks a bit cramped.
An example: 
- Style from the actual spec in google docs: 
- Style from of that specs in specs.canonical.com: 



**3)** In this case this is an issue that comes from the designs in figma that you just applied, but I think it's worth changing the footer style when then side panel is open to have it in a light grey with a border, similar to [Vanilla application layout](https://vanillaframework.io/docs/examples/layouts/application/JAAS), except the footer in only applied to the side panel. The reason I recommend this is because it looks weird or it seems one has reached the bottom of the content. 



Let me know if you have any questions. Thanks!

**Comment by marinacastejon** [comment]
4) The side panel in the figma designs is about half the screen width and in the site it is much wider.

5) In the designs it looks like the whole specs card is clickable. In the developed site we have to click the title. For efficiency of use I would make the whole card clickable. See example in [Charmhub](https://charmhub.io/).

**Comment by anthonydillon** [comment]
> In the designs it looks like the whole specs card is clickable. In the developed site we have to click the title. For efficiency of use I would make the whole card clickable. See example in [Charmhub](https://charmhub.io/).

One con of this approach is that the contents of the card such as authors, etc will not be selectable to copy and paste. Can be frustrating. Also means nothing else on the card can be a link.

**Comment by marinacastejon** [comment]
> > In the designs it looks like the whole specs card is clickable. In the developed site we have to click the title. For efficiency of use I would make the whole card clickable. See example in [Charmhub](https://charmhub.io/).
> 
> One con of this approach is that the contents of the card such as authors, etc will not be selectable to copy and paste. Can be frustrating. Also means nothing else on the card can be a link.

Good points @anthonydillon. Since the remedy is worse, I'd keep the current behaviour then. Thanks

**Comment by marinacastejon** [comment]
Another comment as discussed during stand-up. When the side panel is open, the focus should be trapped inside. When I tested this, the focus when behind the side panel.

**Comment by andesol** [comment]
Closing this because this featured is tackled here: #90

**Pull request closed without merging**